libs/wingsdk/src/target-tf-azure/function.ts

```diff
@@ -158,13 +168,36 @@ export class Function extends cloud.FunctionBase {
     }
   }
 
-  // Adds role to function for given azure scope
+  /**
+   *  Adds role to function for given azure scope
+   *
+   * @param scope - The azure permission scope of the role assignment to create.
+   * @param roleDefinitionName - The name of the role definition to use in the role assignment.
+   */
   public addPermission(scope: string, roleDefinitionName: string) {
     if (!this.permissions) {
       this.permissions = new Map();
     }
-    const existingPermissions = this.permissions.get(scope) ?? [];
-    this.permissions.set(scope, [...existingPermissions, roleDefinitionName]);
+
+    // If the function has already been initialized attach the role assignment directly
+    if (this.function) {
+      if (
+        this.permissions.has(scope) &&
+        this.permissions.get(scope)?.has(roleDefinitionName)
+      ) {
+        return; // already exists
+      }
+
+      new RoleAssignment(this, `RoleAssignment${roleDefinitionName}`, {
```

Comment:
I think if `addPermission("scope1", "perm1")` is called followed by `addPermission("scope2", "perm1")` they will end up with the same name and an error will occur - if so the `scope` should probably be part of the name as well